internal/bson2/raw_document.go

```diff
@@ -17,20 +17,44 @@ package bson2
 import (
 	"encoding/binary"
 	"log/slog"
+	"math"
 
 	"github.com/cristalhq/bson/bsonproto"
 
 	"github.com/FerretDB/FerretDB/internal/types"
-	"github.com/FerretDB/FerretDB/internal/util/debugbuild"
 	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
-// RawDocument represents a BSON document a.k.a object in the binary encoded form.
+// RawDocument represents a singe BSON document a.k.a object in the binary encoded form.
 //
 // It generally references a part of a larger slice, not a copy.
 type RawDocument []byte
 
+// FindRawDocument returns the first BSON document in the byte slice.
+// It should start at the first byte.
+//
+// Returned document might not be valid. It is the caller's responsibility to check it.
+//
+// Use RawDocument(b) conversion instead of b contains exactly one document and no extra bytes.
+func FindRawDocument(b []byte) RawDocument {
+	bl := len(b)
+	if bl < 5 {
+		return nil
+	}
+
+	dl := int(binary.LittleEndian.Uint32(b))
+	if bl < dl {
+		return nil
+	}
+
+	if b[dl-1] != 0 {
```

Comment:
I might be misunderstanding. The last byte of document has to be 0 to be valid `RawDocument`?

Reply:
Yeah, see https://bsonspec.org/spec.html

> `document ::= int32 e_list unsigned_byte(0)`

Reply:
Thanks :+1: